### PR TITLE
Create crud endpoints for trecipe

### PR DESCRIPTION
**Pull request by shizuko-akamoto**
Implemented CRUD endpoints for Trecipe data.
This PR also contains some general server code organization. I went with dividing the code based on model package (ie. `user`, `trecipe`, `destination`, etc). Each model package should have its own corresponding controller, service, interface, and anything else it needs.

## Server Code Structure
**Model file**
Model file should declare mongoose schema and model

**Interface file**
Declares an interface for the model. It seems like a duplicate to have both Mongoose schema and interface, but the reason we need this is so that we can associate typescript type to mongoose model.
For example, for Trecipe model, we associate with Trecipe interface like such:

trecipe.model.ts
`const trecipeModel = mongoose.model<Trecipe & mongoose.Document>('Trecipe', trecipeSchema);`

**Controller file**
In controller, we attach our request handlers to express Router. Every controller should implement the `Controller` interface to ensure it has the path and router fields. Note that the controller should not handle anything with Mongoose model, but rather hand it off to corresponding service to do so.

**Service file**
The service is where we actually interact with Mongoose model to operate on MongoDB.

## Request Vlidation: api.yml and OpenApiValidator

The express generator I used to build server with comes with request/response validation using OpenAPI 3, which I thought was quite useful. If we declare required structure for each request and response in `api.yml` file, `OpenApiValidator` will help us make sure each request and response has the correct required format. If incorrect request/response is encountered, the validator will help us send `400: Bad Request` with appropriate error message.
Example: I sent a post request to `/trecipes` endpoint with empty {} request body, and I get back this error response from the validator.
image

## API Documentation
Another benefit of using OpenAPI 3 is that it can use Swagger UI to document our API. After running the server, if you go to [http://localhost:7000/api-explorer/](http://localhost:7000/api-explorer/), we can access the documentation. 

## Todo
`api.yml` currently contains specs for destinations endpoints, but no endpoint is actually implemented for this PR.

### CODE REVIEW

File: packages/server/server/api/trecipe/trecipe.interface.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+export default interface Trecipe {
+    uuid: string;
+    name: string;
+    description: string;
+    owner: string;
+    isPrivate: boolean;
+    collaborators: Array<string>;
+    image: string;
```
Comment (tianyan-zhu) [inline packages/server/server/api/trecipe/trecipe.interface.ts:13]: Do we need img thumbnail?

File: packages/server/server/api/trecipe/trecipe.model.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+const trecipeSchema = new mongoose.Schema(
+    {
+        uuid: String,
+        name: String,
+        description: String,
+        owner: { type: String, ref: 'User' },
+        isPrivate: Boolean,
+        collaborators: [{ type: String, ref: 'User' }],
+        image: String,
```
Comment (tianyan-zhu) [inline packages/server/server/api/trecipe/trecipe.model.ts:17]: thumbnail?

File: packages/server/server/api/user/user.interface.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export default interface User {
+    username: string;
+    displayName: string;
+    email: string;
+    password: string;
+    trecipes: Array<string>;
```
Comment (tianyan-zhu) [inline packages/server/server/api/user/user.interface.ts:6]: Doesn't user have a profile pic?

File: packages/server/server/api/user/user.model.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import mongoose from 'mongoose';
+import User from './user.interface';
+
+const userSchema = new mongoose.Schema(
+    {
```
Comment (tianyan-zhu) [inline packages/server/server/api/user/user.model.ts:5]: Same here, profile pic?

File: packages/server/server/api/trecipe/trecipe.interface.ts (hunk 1 above)
Comment (shizuko-akamoto) [reply packages/server/server/api/trecipe/trecipe.interface.ts:13]: What's the thumbnail for?? Are you referring to the smaller images on Trecipe cards vs the bigger images in Trecipe page?

File: packages/server/server/api/user/user.interface.ts (hunk 3 above)
Comment (shizuko-akamoto) [reply packages/server/server/api/user/user.interface.ts:6]: Sure, i can add that.

File: packages/server/server/api/destinations/destination.interface.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+export enum DestinationCategory {
+    FOOD = 'food',
+    ACCOMMODATION = 'accommodation',
+    ENTERTAINMENT = 'entertainment',
```
Comment (knox153) [inline packages/server/server/api/destinations/destination.interface.ts:4]: There is a new category? Time to go back to the spreadsheet and reorganize 😅 And it seems like we will be storing multiple categories for each destination, but how do we choose which category icon to display? randomly?

File: packages/server/server/api/destinations/destination.interface.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    uuid: string;
+    name: string;
+    category: DestinationCategory;
+    geometry: {
+        lat: number;
+        lon: number;
+    };
+    formattedAddress: string;
+    formattedPhoneNumber: string;
+    website: string;
```
Comment (knox153) [inline packages/server/server/api/destinations/destination.interface.ts:19]: ohh I forgot to mention one thing, google place detail has no description. So I used the website to replace the description on destination card lol. Dun know what else we could have use there

File: packages/server/server/api/trecipe/trecipe.controller.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        this.router.post(this.path, this.createTrecipe);
+        this.router.get(`${this.path}/:id`, this.getTrecipeById);
+        this.router.delete(`${this.path}/:id`, this.deleteTrecipeById);
+        this.router.put(`${this.path}/:id`, this.updateTrecipeById);
+    }
+
+    private getAllTrecipes(req: Request, res: Response, next: NextFunction) {
+        TrecipeService.getAll().then((trecipes: Array<Trecipe>) => {
+            res.json(trecipes);
+        });
```
Comment (knox153) [inline packages/server/server/api/trecipe/trecipe.controller.ts:28]: Perhaps add error handling

File: packages/server/server/api/destinations/destination.interface.ts (hunk 5 above)
Comment (shizuko-akamoto) [reply packages/server/server/api/destinations/destination.interface.ts:4]: NOooo sorry, my bad. You don't need to change it, we can just stick to what the front end has. We store multiple categories, but I think we can just show arbitrarily the first one in the list for now?

File: packages/server/server/api/destinations/destination.interface.ts (hunk 6 above)
Comment (shizuko-akamoto) [reply packages/server/server/api/destinations/destination.interface.ts:19]: Yea, realized that too. Seems like there's an issue request Google to add description back to their api, but pretty sure it's not gonna be addressed anytime soon. I think I will leave the description field in the model, but we won't display it anywhere I guess LOL.

File: packages/server/server/api/trecipe/trecipe.controller.ts (hunk 7 above)
Comment (shizuko-akamoto) [reply packages/server/server/api/trecipe/trecipe.controller.ts:28]: 😅 will dooo

File: packages/server/server/api/destinations/destination.interface.ts (hunk 5 above)
Comment (knox153) [reply packages/server/server/api/destinations/destination.interface.ts]: okayy, entertainment is a good addition though lol. 
yea as long as it is consistent, we should be good. we dun want users to get different category for the same destination every time they login haha

**Pull request merged**